lib/chef/knife/bootstrap.rb

```diff
@@ -696,9 +696,19 @@ def do_connect(conn_options)
         @connection = TrainConnector.new(host_descriptor, connection_protocol, conn_options)
         connection.connect!
       rescue Train::UserError => e
+        limit ||= 1
         if !conn_options.key?(:pty) && e.reason == :sudo_no_tty
           ui.warn("#{e.message} - trying with pty request")
           conn_options[:pty] = true # ensure we can talk to systems with requiretty set true in sshd config
+          retry
+        elsif config[:use_sudo_password] && (e.reason == :sudo_password_required || e.reason == :sudo_root_password_required) && limit < 3
```

Comment:
Catch the sudo root password related error and retry with password prompt

Reply:
As I can see lots of retries we have added so for definitely need to refactor and generalize code as much as possible